File: NonMaps/src/main/java/com/nonlinearlabs/client/dataModel/value/ClippedValue.java
package com.nonlinearlabs.client.dataModel.value;

class ClippedValue extends Value {

	public double getClippedValue() {
		return clip(getRawValue());
	}

	public double clip(double v) {
		return Math.min(getUpperBorder(), Math.max(v, getLowerBorder()));
	}

	public double getUpperBorder() {
		return 1.0;
	}

	public double getLowerBorder() {
		return isBipolar() ? -1.0 : 0.0;
	}

	@Override
	protected void onRawValueChanged(double oldRawValue, double newRawValue) {
		double oldClippedValue = clip(oldRawValue);
		double newClippedValue = clip(newRawValue);

		if (oldClippedValue != newClippedValue)
			onClippedValueChanged(oldClippedValue, newClippedValue);

		super.onRawValueChanged(oldRawValue, newRawValue);
	}

	protected void onClippedValueChanged(double oldClippedValue, double newClippedValue) {
	}
}

File: projects/epc/playground/src/parameters/value/ClippedValue.cpp
#include "ClippedValue.h"
#include <algorithm>

ClippedValue::ClippedValue(const ScaleConverter *scale, RawValue::tValueType def)
    : super(scale, def)
{
}

ClippedValue::~ClippedValue()
{
}

RawValue::tValueType ClippedValue::getClippedValue() const
{
  return clip(getRawValue());
}

RawValue::tValueType ClippedValue::getClippedValue(RawValue::tValueType v) const
{
  return clip(v);
}

RawValue::tValueType ClippedValue::clip(RawValue::tValueType v) const
{
  return std::min(getUpperBorder(), std::max(v, getLowerBorder()));
}

RawValue::tValueType ClippedValue::getUpperBorder() const
{
  return 1.0;
}

RawValue::tValueType ClippedValue::getLowerBorder() const
{
  return isBiPolar() ? -1.0 : 0.0;
}

void ClippedValue::onRawValueChanged(Initiator initiator, RawValue::tValueType oldRawValue,
                                     RawValue::tValueType newRawValue)
{
  auto oldClippedValue = clip(oldRawValue);
  auto newClippedValue = clip(newRawValue);

  if(oldClippedValue != newClippedValue)
    onClippedValueChanged(initiator, oldClippedValue, newClippedValue);

  super::onRawValueChanged(initiator, oldRawValue, newRawValue);
}
